manager/MaCh3Modes.h

```diff
@@ -0,0 +1,79 @@
+#pragma once
+
+#include <cmath>
+#include <map>
+#include <string>
+
+//MaCh3 includes
+#include "manager/YamlHelper.h"
+#include "manager/Monitor.h"
+#include "manager/MaCh3Exception.h"
+
+/// Enumerator of MaCh3Mode
+using MaCh3Modes_t = int;
+
+/// @brief KS: Class containing information for a single MaCh3Mode
+struct ModeInfo {
+  /// Mode name
+  std::string Name;
+  /// Mode fancy name
+  std::string FancyName;
+  /// Mapping between mode and generator integers
+  std::vector<int> GeneratorMaping;
+  /// @brief KS: Checks MaCh3 modes is associated with a given generator mode
+  inline bool IsMode(const int GenMode) {
+    bool exists = std::find(GeneratorMaping.begin(), GeneratorMaping.end(), GenMode) != GeneratorMaping.end();
+    return exists;
+  }
+};
+
+/// @brief KS: Class describing MaCh3 modes used in the analysis, it is being initialised from config
+class MaCh3Modes {
+ public:
+  /// @brief KS: Initialise MaCh3 modes
+  MaCh3Modes(std::string const &filename);
+  /// @brief KS: Empty destructor
+  virtual ~MaCh3Modes(){};
+
+  /// @brief KS: Print info about initialised modes
+  void Print();
+
+  /// @brief KS: Get number of modes, keep in mind actual number is +1 greater due to unknown category
+  inline int GetNModes(){return NModes;}
+  /// @brief KS: Get mode number based on name, if mode not known you will get UNKNOWN_BAD
+  MaCh3Modes_t GetMode(std::string name);
+  /// @brief KS: Get normal name of mode, if mode not known you will get UNKNOWN_BAD
+  std::string GetMaCh3ModeName(const int Index);
+  /// @brief KS: Get fancy name of mode, if mode not known you will get UNKNOWN_BAD
+  std::string GetMaCh3ModeFancyName(const int Index);
+  /// @brief KS: Get MaCh3 mode from generator mode
+  MaCh3Modes_t GetModeFromGenerator(const int Index);
+  /// @brief Get class name
+  inline std::string GetName()const {return "MaCh3Modes";};
+
+ private:
+  /// @brief KS: Make sure we don't have two modes with the same name
+  inline MaCh3Modes_t EnsureModeNameRegistered(std::string const &name);
+
+  /// @brief KS: Add new mode
+  inline void DeclareNewMode(std::string const &name,
+                      std::string const &fancyname,
+                      std::vector<int> const &GenMap);
+
+  /// @brief KS: Fill ModeMap
+  inline void PrepareMap();
+
+  /// KS: Handy map which halpes find mode number based on string
+  std::map<std::string, MaCh3Modes_t> Mode;
+  /// KS: Main map stroing info about used modes
+  std::map<MaCh3Modes_t, ModeInfo> fMode;
```

Comment:
Are we happy with MaCh3Modes_t? I am just wondering whether practically it will just confuse people? If it is a std::map<int, ModeInfo> all the functionality will be the same right?

Reply:
My idea behind MaCh3Modes_t is to sort of imilate enum.

Previolsy you had MaCh3Mode kCCQE;
Now you have MaCh3Modes_t  kCCQE = ...

From compiler point of view MaCh3Modes_t   is just an `int`, but I think (I can be wrong) MaCh3Modes_t  helps read code. As it indicate that you are dealing with pseudo enum, not just normal int

Reply:
I think the readibility is improved too but I got a little bit caught out when looking at it to begin with.